bicytok/distanceMetricFuncs.py

```diff
@@ -1,641 +1,221 @@
-        ax.set_xticks(range(len(receptor_names)))
-        ax.set_xticklabels(receptor_names, rotation="vertical")
-
-    return sorted_results
-
-
-def EMD_3D(dataset1, target_cells, ax=None):
+
+    KL_div_vals = np.full(recAbundances.shape[1], np.nan)
+    EMD_vals = np.full(recAbundances.shape[1], np.nan)
+
+    for rec in range(recAbundances.shape[1]):
+        targAbun, offTargAbun = normalize(recAbundances[:, rec], targ, offTarg)
+
+        if np.mean(targAbun) > np.mean(offTargAbun):
+            KL_div_vals[rec] = calculateKLDiv(targAbun, offTargAbun, dims=1)
+            EMD_vals[rec] = stats.wasserstein_distance(
+                targAbun, offTargAbun
+            )  # Consider switching/comparing to ot.emd2_1d
+
+    return KL_div_vals, EMD_vals
+
+
+def KL_EMD_2D(recAbundances, targ, offTarg):
     """
-    returns list of descending EMD values for specified target cell (3 receptors)
-    returns list of descending EMD values for specified target cell (3 receptors)
-    :param dataset1: DataFrame containing the dataset
-    :param target_cells: Target cell type for analysis
-    :param ax: Matplotlib Axes object for plotting (optional)
+    Calculates 2D EMD and KL Divergence between the target and off-target populations of two receptors, across multiple receptors
+    :param recAbundances: an array of receptor abundances across cells (rows) and receptors (columns)
+    :param targ: a binary/boolian vector indicating indices of target cells
+    :param offTarg: a binary/boolian vector indicating indices of off-target cells
     :return:
-    List of tuples (format: (recep1, recep2, recep 3, OT value) containing optimal transport distances and receptor information for 3D analysis
+        KL_div_vals: an array of KL Divergences between the on- and off-target abundances of multiple receptors
+            for each entry the value is calculated for two receptors, so every entry is for a different receptor pair
+            the array is triangular because it is symmetric across the diagonal
+            ???the diagonal values are the 1D distances??? Check this
+        EMD_vals: similar to KL_div_vals but with EMDs
+    """
+
+    # Do conversion factors in functions?
+    # weightDF = convFactCalc(recAbundances) ...
+
+    KL_div_vals = np.empty([recAbundances.shape[1], recAbundances.shape[1]])
+    KL_div_vals[:] = np.nan
+    EMD_vals = np.empty([recAbundances.shape[1], recAbundances.shape[1]])
+    EMD_vals[:] = np.nan
+
+    for rec1 in range(recAbundances.shape[1]):
+        targAbun1, offTargAbun1 = normalize(recAbundances[:, rec1], targ, offTarg)
+
+        # for rec2 in range(recAbundances.shape[1]): #Make square matrix outputs (symmetrical across diagonal)
+        for rec2 in range(rec1 + 1):  # Make triangular matrix outputs
+            targAbun2, offTargAbun2 = normalize(recAbundances[:, rec2], targ, offTarg)
+
+            # How to interpret case when rec1==rec2?
+            # -Could result in 1D distances
```

Comment:
I'd fill in the 1D differences. This will make it easier to have these available as a basis of comparison later.